src/contracts/periphery/RolesAuthority.sol

```diff
@@ -0,0 +1,97 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity >=0.8.0;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+/// @notice A roles based access control contract that supports up to 256 roles.
+/// @author Modified from Solmate (https://github.com/transmissions11/solmate/blob/main/src/auth/authorities/RolesAuthority.sol)
+
+contract RolesAuthority is Ownable {
+    /*//////////////////////////////////////////////////////////////
+                                 EVENTS
+    //////////////////////////////////////////////////////////////*/
+
+    event UserRoleUpdated(address indexed user, uint8 indexed role, bool enabled);
+
+    event RoleCapabilityUpdated(uint8 indexed role, address indexed target, bytes4 indexed functionSig, bool enabled);
+
+    /*//////////////////////////////////////////////////////////////
+                               CONSTRUCTOR
+    //////////////////////////////////////////////////////////////*/
+
+    constructor(address _owner) {
+        _transferOwnership(_owner);
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                            ROLE/USER STORAGE
+    //////////////////////////////////////////////////////////////*/
+
+    mapping(address => bytes32) public getUserRoles;
+
+    mapping(address => mapping(bytes4 => bool)) public isCapabilityPublic;
+
+    mapping(address => mapping(bytes4 => bytes32)) public getRolesWithCapability;
+
+    function doesUserHaveRole(address user, uint8 role) public view virtual returns (bool) {
+        return (uint256(getUserRoles[user]) >> role) & 1 != 0;
+    }
+
+    function doesRoleHaveCapability(
+        uint8 role,
+        address target,
+        bytes4 functionSig
+    ) public view virtual returns (bool) {
+        return (uint256(getRolesWithCapability[target][functionSig]) >> role) & 1 != 0;
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                           AUTHORIZATION LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    function canCall(
+        address user,
+        address target,
+        bytes4 functionSig
+    ) public view virtual returns (bool) {
+        return
+            bytes32(0) != getUserRoles[user] & getRolesWithCapability[target][functionSig];
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                   ROLE CAPABILITY CONFIGURATION LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    function setRoleCapability(
+        uint8 role,
+        address target,
+        bytes4 functionSig,
+        bool enabled
+    ) public virtual onlyOwner {
+        if (enabled) {
+            getRolesWithCapability[target][functionSig] |= bytes32(1 << role);
+        } else {
+            getRolesWithCapability[target][functionSig] &= ~bytes32(1 << role);
+        }
+
+        emit RoleCapabilityUpdated(role, target, functionSig, enabled);
+    }
```

Comment:
This goes much further in enshrining RBAC than uniform delegation. Having the additional role construct is now beginning to enter the problem space of account management. I'm not certain that is the goal here.

Reply:
Are roles the wrong abstraction here?

Reply:
Synced with Z -> should be non roles based